rir/src/interpreter/interp.cpp

```diff
@@ -24,6 +22,112 @@ extern Rboolean R_Visible;
 
 using namespace rir;
 
+struct CallContext;
+struct FinalCallContext;
+struct FinalDispatchContext;
+
+struct DispatchContext {
+    R_bcstack_t* eagerArgs;
+    Function* caller;
+    SEXP callerEnv;
+    CallSite* callSite;
+    unsigned nargs;
+    SEXP ast;
+
+    DispatchContext(Code* c, unsigned id, R_bcstack_t* eagerArgs,
+                    unsigned nargs, SEXP callerEnv, Context* ctx)
+        : eagerArgs(eagerArgs), caller(c->function()), callerEnv(callerEnv),
+          callSite(c->callSite(id)), nargs(nargs),
+          ast(cp_pool_at(ctx, callSite->call)) {}
+
+    DispatchContext(Code* c, unsigned id, unsigned nargs, SEXP callerEnv,
+                    Context* ctx)
+        : DispatchContext(c, id, nullptr, nargs, callerEnv, ctx) {}
+
+    DispatchContext(R_bcstack_t* eagerArgs, Function* caller, SEXP callerEnv,
+                    CallSite* callSite, unsigned nargs, SEXP ast)
+        : eagerArgs(eagerArgs), caller(caller), callerEnv(callerEnv),
+          callSite(callSite), nargs(nargs), ast(ast) {}
+
+    bool hasEagerArgs() const { return eagerArgs; }
+
+    Code* arg(unsigned i) const {
+        assert(i < nargs);
+        return caller->codeAt(callSite->args()[i]);
+    }
+
+    SEXP eagerArg(unsigned i) const {
+        assert(eagerArgs && i < nargs);
+        return ostack_at_cell(eagerArgs + i);
+    }
+
+    SEXP getFixedupAst() const;
+    DispatchContext fixupAst() const {
```

Comment:
yes, you are right, but it should not be virtual. In my opinion `virtual` should only be used if you actually do have a call site where it is not statically clear, what should be called, which is not the case here. Also, the idea with this struct is that it is basically just a collection of things that would otherwise have to be passed as 5 separate arguments to all those functions. So we are relying on the compiler here, to make passing the struct as efficient, as passing the arguments separately. This will certainly not be the case, if the struct has virtual methods.

Reply:
> But if we use pointers or references, then yes, the base version would be called. The only problem would be if someone passed a CallContext into doDispatch(DispatchContext&,...)

even then it is not a problem, since both do the same. You'd just get a FinalDispatchContext, which means you loose access to the callee, but that's it...

Reply:
My concern is that there's too many gotchas here (thanks C++). Even my example of demonstrating a gotcha fell victim to object slicing!

If it's intentionally *not* `virtual`, then I would have liked to see some comment, since the usual pattern is to make it `virtual`. But there's still places where it can go wrong, and I'd prefer it if we could enforce it in the compiler.

If the point is to just pass things around in a struct, would it make more sense to use composition instead of inheritance? Does the code rely on the fact that `CallContext` is a subtype of `DispatchContext`, or can we work around that?